### PR TITLE
fix(windows): OSK toolbar sync

### PR DESCRIPTION
**Pull request by mcdurdin**
As Keyman now monitors the Windows keyboard list from Control Panel, it is no longer sufficient to refresh the toolbar only when a Keyman keyboard is updated, as at that point Windows has not refreshed its own language list in the registry. Instead, we need to monitor changes to the relevant registry key and then refresh when those changes are 'complete'.

There is a problem: there does not appear to be a good way to monitor for changes to Windows language settings. `WM_SETTINGCHANGE` is sent when a language is added by the user in Settings, but the actual change happens asynchronously, and we are not notified when the changes are complete.

We also do not get notifications when languages are added via APIs.

Given the async nature of the language profile changes, this leads to race conditions when we try to refresh the OSK toolbar. My preferred solution is to monitor the relevant key in a separate thread, and when changes are received we could refresh every 125msec for 5 seconds after the last change is detected. It's a bit yuck perf-wise but would probably be pretty transparent to the user. (We reset that poll timer every time a change is detected)

In testing, this looks good. It means the changes appear as soon as ready, without substantial delays or performance issues. Polling this makes me feel a bit dirty, but I'll just take a shower and be okay.

# User Testing

@MakaraSok can you try enabling/disabling keyboards, adding/removing keyboards and/or languages, and monitor the OSK toolbar to see if it reflects the changes reasonably quickly?

### CODE REVIEW

File: windows/src/engine/keyman/UfrmKeyman7Main.pas [hunk 1]
```diff
@@ -528,19 +532,8 @@ procedure TfrmKeyman7Main.FormCreate(Sender: TObject);
...
-  wm_keyman_control := RegisterWindowMessage('WM_KEYMAN_CONTROL');
-  wm_keyman_control_internal := RegisterWindowMessage('WM_KEYMAN_CONTROL_INTERNAL');   // I3933
   sMsg_TaskbarRestart := RegisterWindowMessage('TaskbarCreated');
-
-  ChangeWindowMessageFilter(wm_keyman_control, MSGFLT_ADD);
-  ChangeWindowMessageFilter(wm_keyman_globalswitch, MSGFLT_ADD);
-  ChangeWindowMessageFilter(wm_keyman_globalswitch_process, MSGFLT_ADD);
-  ChangeWindowMessageFilter(wm_keyman_control_internal, MSGFLT_ADD);   // I3933
   ChangeWindowMessageFilter(sMsg_TaskbarRestart, MSGFLT_ADD);
-  ChangeWindowMessageFilter(WM_USER_PlatformComms, MSGFLT_ADD);
```
Comment (mcdurdin) [inline windows/src/engine/keyman/UfrmKeyman7Main.pas:543]: Most of these moved to initialization section so they will be ready even before `FormCreate` (as `CreateWnd` can be called before `FormCreate` I believe)

**Review by jahorton — APPROVED** [review]

**Comment by MakaraSok** [comment]
Most of the time, the toolbar on the OSK synced with the keyboard installed and option chosen in the Configuration.

One thing is when adding a language to the existing keyboard, i.e. add 'cmo' to Khmer Angkor keyboard would trigger the message below:

image

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.215-beta](https://keyman.com/downloads/releases/beta/14.0.215)

**Comment by mcdurdin** [comment]
> One thing is when adding a language to the existing keyboard, i.e. add 'cmo' to Khmer Angkor keyboard would trigger the message

That message is to be expected if you already have 4 transient languages installed. It cam eout as a result of the changes in #4157.

**Comment by MakaraSok** [comment]
What is interesting is that even if I have only two keyboards/languages installed, this still pops up. 

Do we have a documentation somewhere about "transient language"?

**Comment by MakaraSok** [comment]
I could be completely wrong, but let's say I'd like to add 'cmo' language to the Khmer Angkor keyboard, now in the process the message appears. The screenshot below shows only one keyboard is installed.

image

**Comment by mcdurdin** [comment]
Check your Windows language settings; there may be left over Windows languages installed which are consuming the available transient languages even if Keyman isn't using them.

**Comment by mcdurdin** [comment]
> Do we have a documentation somewhere about "transient language"?

Not really at this stage. A TODO perhaps?

**Comment by MakaraSok** [comment]
Only one keyboard is installed and two languages are in the Windows languages. But when 'cmo' is added to the Khmer Angkor, that same message popped up. Now, instead of 'cmo' gets added to the keyboard, Khmer (NiDA) is crept in.

image

In anyway, there is no more than four keyboards/languages that I can see in sight, unless they are not at where I am looking.

**Comment by mcdurdin** [comment]
@MakaraSok, can you do the following steps for checking the installed Windows languages:

1. Start Powershell (<kbd>Win</kbd>+<kbd>X</kbd>, select **Windows Powershell** may be the fastest way)
2. Type the following command:

```
get-winuserlanguagelist
```

3. Paste the result into a comment here.

**Comment by MakaraSok** [comment]
Here is the result:

```
LanguageTag     : en-US
Autonym         : English (United States)
EnglishName     : English
LocalizedName   : English (United States)
ScriptName      : Latin
InputMethodTips : {0409:00000409, 0409:{FE0420F1-38D1-4B4C-96BF-E7E20A74CFB7}{0645B242-CC7C-4316-B64E-8974B82634DF}}
Spellchecking   : True
Handwriting     : False

LanguageTag     : km
Autonym         : ភាសាខ្មែរ
EnglishName     : Khmer
LocalizedName   : Khmer
ScriptName      : Khmer
InputMethodTips : {0453:{FE0420F1-38D1-4B4C-96BF-E7E20A74CFB7}{7099BD1D-4E1F-4627-828F-372D0E7DB869}}
Spellchecking   : True
Handwriting     : True
```

**Comment by mcdurdin** [comment]
Here's another test to run. Run the following command in a command prompt and paste the results here:

```
reg query "hkcu\control panel\international\user profile" /s
```

**Comment by MakaraSok** [comment]
```
PS C:\Users\makara.PHNOMPENH> reg query "hkcu\control panel\international\user profile" /s

HKEY_CURRENT_USER\control panel\international\user profile
    Languages    REG_MULTI_SZ    en-US\0km
    ShowAutoCorrection    REG_DWORD    0x1
    ShowTextPrediction    REG_DWORD    0x1
    ShowCasing    REG_DWORD    0x1
    ShowShiftLock    REG_DWORD    0x1

HKEY_CURRENT_USER\control panel\international\user profile\azj-Cyrl-AM
    CachedLanguageName    REG_SZ    North Azerbaijani (azj-Cyrl-AM)

HKEY_CURRENT_USER\control panel\international\user profile\daq-Deva-IN
    CachedLanguageName    REG_SZ    Dandami Maria (daq-Deva-IN)

HKEY_CURRENT_USER\control panel\international\user profile\el-Cyrl-UA
    CachedLanguageName    REG_SZ    Modern Greek (1453-) (el-Cyrl-UA)

HKEY_CURRENT_USER\control panel\international\user profile\emu-Deva
    CachedLanguageName    REG_SZ    Eastern Muria (emu-Deva)

HKEY_CURRENT_USER\control panel\international\user profile\en-US
    CachedLanguageName    REG_SZ    @Winlangdb.dll,-1121
    0409:00000409    REG_DWORD    0x1
    0409:{FE0420F1-38D1-4B4C-96BF-E7E20A74CFB7}{9D21FFC2-0C18-4FF4-910B-83C510F87692}    REG_DWORD    0x2

HKEY_CURRENT_USER\control panel\international\user profile\esg-Deva-IN
    CachedLanguageName    REG_SZ    Aheri Gondi (esg-Deva-IN)

HKEY_CURRENT_USER\control panel\international\user profile\fmu-Deva
    CachedLanguageName    REG_SZ    Far Western Muria (fmu-Deva)

HKEY_CURRENT_USER\control panel\international\user profile\gno-Deva
    CachedLanguageName    REG_SZ    Gondi (Northern) (gno-Deva)

HKEY_CURRENT_USER\control panel\international\user profile\gno-Deva-IN
    CachedLanguageName    REG_SZ    Gondi (Northern) (gno-Deva-IN)

HKEY_CURRENT_USER\control panel\international\user profile\got-Goth
    CachedLanguageName    REG_SZ    @Winlangdb.dll,-1191

HKEY_CURRENT_USER\control panel\international\user profile\grc-Linb-GR
    CachedLanguageName    REG_SZ    Ancient Greek (to 1453) (grc-Linb-GR)

HKEY_CURRENT_USER\control panel\international\user profile\he-Brai-IL
    CachedLanguageName    REG_SZ    Hebrew (he-Brai-IL)

HKEY_CURRENT_USER\control panel\international\user profile\khk-Tibt-CN
    CachedLanguageName    REG_SZ    Halh Mongolian (khk-Tibt-CN)

HKEY_CURRENT_USER\control panel\international\user profile\kjg-Thai-TH
    CachedLanguageName    REG_SZ    Khmu (kjg-Thai-TH)

HKEY_CURRENT_USER\control panel\international\user profile\km
    CachedLanguageName    REG_SZ    @Winlangdb.dll,-1251
    0453:{FE0420F1-38D1-4B4C-96BF-E7E20A74CFB7}{BB0CA12F-6820-49F2-B544-0C65C8EA39BD}    REG_DWORD    0x1

HKEY_CURRENT_USER\control panel\international\user profile\mrr-Deva-IN
    CachedLanguageName    REG_SZ    Maria (India) (mrr-Deva-IN)

HKEY_CURRENT_USER\control panel\international\user profile\mut-Deva-IN
    CachedLanguageName    REG_SZ    Western Muria (mut-Deva-IN)

HKEY_CURRENT_USER\control panel\international\user profile\oru-Arab-PK
    CachedLanguageName    REG_SZ    Ormuri (oru-Arab-PK)

HKEY_CURRENT_USER\control panel\international\user profile\wbl-Latn-PK
    CachedLanguageName    REG_SZ    Wakhi (wbl-Latn-PK)

HKEY_CURRENT_USER\control panel\international\user profile\wsg-Telu-IN
    CachedLanguageName    REG_SZ    Adilabad Gondi (wsg-Telu-IN)

HKEY_CURRENT_USER\control panel\international\user profile\ydg-Arab-PK
    CachedLanguageName    REG_SZ    Yidgha (ydg-Arab-PK)
```

And here is how the Keyboard Layouts tab and Keyman Menu look:
image

**Comment by mcdurdin** [comment]
Do you still get the same problem updating now? Or has that issue gone away?

**Comment by MakaraSok** [comment]
Yes, the issue has not diverged. OSK toolbar is alright, but adding a language to a keyboard causes this:
image

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)

**Pull request merged**